app/migrate_module_account.go

```diff
@@ -0,0 +1,78 @@
+package app
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
+// MigratePreInitializedModuleAccounts migrates module accounts that were pre-initialized as BaseAccounts to ModuleAccounts,
+// or creates new module accounts if they don't exist.
+func MigratePreInitializedModuleAccounts(
+	ctx sdk.Context,
+	ak authkeeper.AccountKeeper,
+	moduleAccountsToInitialize []string,
+) error {
+	for _, module := range moduleAccountsToInitialize {
+		addr, perms := ak.GetModuleAddressAndPermissions(module)
+		if addr == nil {
+			return fmt.Errorf(
+				"failed to get module address and permissions for module %s",
+				module,
+			)
+		}
+
+		acc := ak.GetAccount(ctx, addr)
+		// The account has not been initialized yet
+		if acc == nil {
+			initModuleAccount(ctx, ak, module, perms...)
+			continue
+		}
+
+		_, isModuleAccount := acc.(authtypes.ModuleAccountI)
+		if isModuleAccount {
+			ctx.Logger().Info(fmt.Sprintf(
+				"account for module %s was correctly initialized, skipping",
+				module,
+			))
+			continue
+		}
+
+		// Migrate from base account to module account
+		baseAccount, ok := acc.(*authtypes.BaseAccount)
```

Comment:
are `ModuleAccountI` and `BaseAccount` the only options?

Reply:
it has [two](https://github.com/cosmos/cosmos-sdk/blob/v0.45.16/x/auth/types/account.go#L21-L25) BaseAccount and ModuleAccount structs and various interfaces on top